chalk-solve/src/clauses/builtin_traits.rs

```diff
@@ -32,12 +33,34 @@ pub fn add_builtin_program_clauses<I: Interner>(
             WellKnownTrait::Sized => sized::add_sized_program_clauses(db, builder, &trait_ref, ty),
             WellKnownTrait::Copy => copy::add_copy_program_clauses(db, builder, &trait_ref, ty),
             WellKnownTrait::Clone => clone::add_clone_program_clauses(db, builder, &trait_ref, ty),
+            WellKnownTrait::FnOnce | WellKnownTrait::FnMut | WellKnownTrait::Fn => {
+                fn_family::add_fn_trait_program_clauses(db, builder, trait_ref.trait_id, self_ty)
+            }
             // Drop impls are provided explicitly
             WellKnownTrait::Drop => (),
         }
     });
 }
 
+/// Like `add_builtin_program_clauses`, but for `DomainGoal::Normalize` involving
+/// a projection (e.g. `<fn(u8) as FnOnce<(u8,)>>::Output`)
+pub fn add_builtin_assoc_program_clauses<I: Interner>(
+    db: &dyn RustIrDatabase<I>,
+    builder: &mut ClauseBuilder<'_, I>,
+    well_known: WellKnownTrait,
+    proj: &ProjectionTy<I>,
+) {
+    match well_known {
+        WellKnownTrait::FnOnce => {
+            let interner = db.interner();
+            let self_ty = AliasTy::Projection(proj.clone()).self_type_parameter(interner);
```

Comment:
Nit: This wrapping could be removed by just passing `self_ty` to *this* function from `clauses` or by passing the `AliasTy` from `clauses`